### ISSUE TITLE
RFE: sign file signatures

### ISSUE THREAD

**Issue by pmatilai** [rpm-software-management/rpm#2224]
File signatures (IMA, fsverify) are quite unlike the package signatures, but all of them live in the signature header unprotected. This means that even on a signed package, somebody can strip out and/or replace file signatures without rpm even knowing it (even though the kernel presumably would reject unknown signatures), which seems kinda unacceptable. The file signatures themselves should be signed by signing a package, which implies an order as well: file signatures cannot be added to an already signed package. Seems kinda obvious when written down.

The file signatures cannot go into the main header because the main header is considered immutable after a build finishes: adding file signatures does not affect the file content or the functionality on a system where file signature technologies are not enforced, so it does not make sense to consider a file-signed package a new package. So it needs to be something else, such as including the file signatures in the hashed data for the package signature, adding a new signature type to the signature header (both caveats of their own), or some other means.

**Comment by pmatilai** [comment]
Been thinking about this on and off. While ideally a signature is something you can verify without parsing the signed contents first, we'll always need to parse the signature header no matter what [*]. So I've been pondering something along the lines of

- collect the tags numbers to be signed into an int32 array in the signature header
- hash that array, and then the tag intros and the actual data in the order specified by the array
- ...and sign/verify the result

It's a relatively complicated procedure but AFAICS it would allow signing the signature header contents in a way that is still verifiable post-install too (think rpm -V). There are no doubt other ways to sign+verify the signature header itself (mask select parts with zeros or something), but those wouldn't work post-install where we no longer have the signature header available as a whole.

@nwalfield @simo5 @DemiMarie - does that seem totally crazy? :sweat_smile: 

Of course now that we support multiple signatures, we need'd to do the same for the signature-signature too. Ie anytime we add a signature, we'd add one for the signature header and one for the header, in a similar structure.

[*] ignoring external signatures: it's not a solution, it's just a different set of problems

**Comment by simo5** [comment]
I do not think we want to get into the rabbit hole of considering multiple signatures this way.
If people really want to bind multiple signatures together they will be able to issue a hybrid signature in future that does the proper binding at the signature level.

If you still wanted to sign headers then the proper solution would require multistage signing:
1) you would need a way to tell rpm what signatures you are planning to use, so that you can lay out the headers with all the data save the actual signature bytes, those would be skipped or set to an array of zeros.
2) you would then apply each signature in turn, and finally you would fill the zeroed arrays with the signatures produced. Verification would need to zero out all signature fields before feeding the data to the verification function.

I think this is needlessly complicated and will prevent adding additional features by 3rd parties a feature I think has significant promise.

I would rather leave it to local policies to enforce specific signature types and specific keys to be validated and not try to create this kind of binding.

**Comment by simo5** [comment]
Just to be clear I do not consider it reasonable to apply signatures to the content and then again additional signatures just to the headers, that would create a matrix explosion of possibilities rather quickly, and render both policy and testing needlessly complicated.

**Comment by DemiMarie** [comment]
> Been thinking about this on and off. While ideally a signature is something you can verify without parsing the signed contents first, we'll always need to parse the signature header no matter what [*]. So I've been pondering something along the lines of
> 
> * collect the tags numbers to be signed into an int32 array in the signature header
> * hash that array, and then the actual tag intros and data for the tags in the array, in the order specified by that array
> * ...and sign/verify the result
> 
> It's a relatively complicated procedure but AFAICS it would allow signing the signature header contents in a way that is still verifiable post-install too (think rpm -V). There are no doubt other ways to sign+verify the signature header itself (mask select parts with zeros or something), but those wouldn't work post-install where we no longer have the signature header available as a whole.
> 
> @nwalfield @simo5 @DemiMarie - does that seem totally crazy? 😅
> 
> Of course now that we support multiple signatures, we need'd to do the same for the signature-signature too. Ie anytime we add a signature, we'd add one for the signature header and one for the header, in a similar structure.
> 
> [*] ignoring external signatures: it's not a solution, it's just a different set of problems

It’s better than nothing, but I would prefer if there was a single signature of both main and signature headers and if all of the entries in the signature header (other than the signature entry itself) had to be signed.  RPMv6 is not released yet so there are no backwards compatibility concerns here.

From a cryptographic perspective, the best solution would be for the signature header to be _only_ signatures on the rest of the data, and for e.g. IMA and fsverity signing to just recreate the main header.  Since the IMA and fsverity signatures are part of what gets installed, they really ought to be treated as part of the package, not as something separate.

> If you still wanted to sign headers then the proper solution would require multistage signing:
> 
> 1. you would need a way to tell rpm what signatures you are planning to use, so that you can lay out the headers with all the data save the actual signature bytes, those would be skipped or set to an array of zeros.
> 2. you would then apply each signature in turn, and finally you would fill the zeroed arrays with the signatures produced. Verification would need to zero out all signature fields before feeding the data to the verification function.

Multistage signing is definitely the best way to go if possible.

> I think this is needlessly complicated and will prevent adding additional features by 3rd parties a feature I think has significant promise.

It’s not needlessly complicated IMO.  What do you mean by “will prevent adding additional features by 3rd parties”?  Packages should not be able to be altered with after signing without invalidating the signature; that’s the whole point of digital signatures.

**Comment by DemiMarie** [comment]
I see.  In that case multi-stage signing would not work, but RPM headers are non-malleable once the constraints I mentioned above are enforced, so one can sign “every header but this one” in an unambiguous and secure way.  I think this is @pmatilai’s proposal.

**Comment by simo5** [comment]
If the header with the signatures is excluded all that needs to be defined is only what else and in which order needs to be signed, and I think a single signature covering all is just fine.

**Comment by pmatilai** [comment]
Hmm, perhaps I need to clarify this: I don't mean signing the other OpenPGP signatures, but the other content in the so called signature header: digests and arrays of file-level signatures which are "free-standing" at the moment: you can modify them without rpm noticing at all. I suppose you can't install file signatures the kernel doesn't trust, but you could eg remove signatures from some files, allegedly to some consequences. It seems a little fishy at best.

**Comment by DemiMarie** [comment]
> Also what I'm talking about here would be a second signature alongside each "normal" signature we add. It needs to be separate because we need to be able to verify the main header independently for post-install signature verification, that doesn't change with V6 at all. What we can do in the V6 space is actually quite limited because of compatibility requirements.

This can be avoided by not actually signing the main header.  Instead, hash the main header,  hash the data from the signature header, and concatenate the hashes.  Then sign the concatenated hashes.  The concatenated hashes can be stored in the rpmdb for use by verification.

**Comment by pmatilai** [comment]
Hmm, concatenate as in just slop one at the end of another? Like:

```
  717cf2bbc006701d1993be3de4687ed8687d3124
+ 20fef0b9bc5bc8236e757dfb42d5cf16f13ce75a
= 717cf2bbc006701d1993be3de4687ed8687d312420fef0b9bc5bc8236e757dfb42d5cf16f13ce75a
```

The relevant signature data is copied to the end of the the main header, and so we can recalculate the per-tag hashes from there (just need to ignore the offset of the tag when calculating the hash). And since the main header can be hashed as-is... we should be able to verify the signature post-install too. 

So that should indeed avoid the need for maintaining a second set of signatures. OTOH this is a lot more complicated to calculate (and implement, including 3rd party signing-servers which there seem to be around every corner nowadays) than the traditional signatures, which we'll still need to support for compatibility anyway.

**Comment by DemiMarie** [comment]
Concrete proposal:

- New-style signatures are _required_ in v6 packages.
- New-style signatures always cover both headers, including the entire main header and all of the signature header except for the tag data entry containing the new-style signatures.
- If a new-style signature is present, the signature header must be a single contiguous region and its entries must be sorted.
- To compute the hash of the signature header:
  - Hash the first 8 bytes.
  - Hash all of the tag data entries until the one for the new-style signature is reached.  If the tag data entries are not sorted by increasing tag, the package is malformed.
  - If there are any remaining tag data entries:
    - Create a copy of the remaining tag data entries.
    - If the remaining tag data entries are not sorted by increasing tag, the package is malformed.
    - Subtract the length (in bytes) of the tag data entry for the new-style signature, rounded up to the next multiple of 8, from the start position of the remaining tag data entries.
    - Hash the remaining tag data entries.
  - Hash the first byte after the tag data entries (inclusive) until (but not including) the first byte of the tag data entry for the new-style signature.
  - Hash all data in the signature header after the new-style signature.

> It is not the best way to go as it may let implementations forget to check the hashes after the signature checking but it is an option to consider if it is simpler.

My thought was that the hash of the signature header would be computed during installation and stored in the rpmdb.  It would not be included in packages.

**Comment by pmatilai** [comment]
We'll need to accommodate v4 signatures in v6 packages for compatibility. So the hashing would need to selectively skip over all the OpenPGP signature items in the header, including legacy ones, and since it needs to do it tag by tag anyhow, doing something else (copying and whatnot) for the what comes after is just an extra complication with little benefit.

We'll also need to accommodate for future expansion too: we need to preserve the ability to add some other signature types later (be it v7 or whatever) without breaking the scheme. So there's another reason we can't just hash whatever comes after. This is basically why my initial idea includes a that int32 type array that contains all the tags the verifier needs to hash. The alternative is to just hardcode a list of tags that this version of rpm will hash, and ignore whatever else may be there.

Also, just as a reality check: what we can actually do in the time left before 6.0 alpha is very limited at point, so whether any of this happens in time for that, I don't know. But this is a useful discussion to have because there's always the next version...

**Comment by pmatilai** [comment]
> This can be avoided by not actually signing the main header. Instead, hash the main header, hash the data from the signature header, and concatenate the hashes. Then sign the concatenated hashes. 

Signing just the hash(es) kinda seems like an attractive proposal at first, but ... AFAICS this means the hash becomes a single point of failure: if the hash is broken then *all* the signatures on it are broken at once, no matter what *their* algorithm.

> New-style signatures are required in v6 packages.
> If a new-style signature is present, the signature header must be a single contiguous region and its entries must be sorted.

That's an interesting point of view, something I've been kinda approaching with the automatic signing on build (with a relatively low-value key) + enforced checking by default: the idea that there are no unsigned packages circulating around. But if the signature is simply a mandatory part of the format, and rpmsign refuses to delete the last signature (ie you can only replace it with another)... :thinking:  There might be something there. It's not quite that simple though, we're expecting to backport the multisignature support to 4.x so v6 packages wouldn't be the only ones carrying those, so the presence of such a signature cannot be the only indication of the format and v4 packages may have all these wrinkles (see eg #3469)...

**Comment by nwalfield** [comment]
> > This can be avoided by not actually signing the main header. Instead, hash the main header, hash the data from the signature header, and concatenate the hashes. Then sign the concatenated hashes.
> 
> Signing just the hash(es) kinda seems like an attractive proposal at first, but ... AFAICS this means the hash becomes a single point of failure: if the hash is broken then _all_ the signatures on it are broken at once, no matter what _their_ algorithm.

This is true anyway.  No real scheme uses asymmetric crypto to sign the actual data.  The data is hashed, and the hash is what is signed.  If you double hash, you should use the same hash algorithm.

**Comment by pmatilai** [comment]
The data is hashed of course yes, but the hash algorithm is (or at least can be) specific to each signature.

**Comment by pmatilai** [comment]
Okay - you could have an array of hashes for different algorithms (ie if a new signature uses a hash that's not already there, add it). But that gets increasingly complex and fragile in other ways.

**Comment by nwalfield** [comment]
> but the hash algorithm is (or at least can be) specific to each signature.

Yes.  But for a given signature, choose one hash and use that everywhere.  Otherwise the security is reduced to the minimum of the hash functions that you use.

**Comment by pmatilai** [comment]
Yup - that's exactly my point: we can now have arbitrary number of signatures with different algorithms each, stored in a single array. If we signed hashes instead, we'd have to store those hashes (separately) for each, to permit the same flexibility (and safety) in variety. Which is adds complexity, but of course its all a tradeoff.

**Comment by simo5** [comment]
Sounds like something that should not be added hastily.
Is he initial description of "protecting" file signatures the only reason to do this?
What is the attack vector here?

**Comment by DemiMarie** [comment]
For one, the Linux kernel parser for IMA or fsverity signatures could have a security vulnerability, as could RPM’s own handling of such signatures.
